File: 1020-number-of-enclaves/1020-number-of-enclaves.cpp
class Solution {
public:
    int DR[4]={0,0,1,-1};
    int DC[4]={1,-1,0,0};
    void dfs(int i,int j,vector<vector<int>>& grid,int &count,bool removal){
        int row=grid.size();
        int col=grid[0].size();
        if( i<0 || j<0 || i>row-1 || j>col-1 || grid[i][j]!=1) return;
        // mark visited nodes with -1
        grid[i][j]=0;
        dfs(i+1,j,grid,count,removal);
        dfs(i-1,j,grid,count,removal);
        dfs(i,j+1,grid,count,removal);
        dfs(i,j-1,grid,count,removal);
        if(removal) count++; 
    }
    
    int numEnclaves(vector<vector<int>>& grid) {
        int row = grid.size(),col = grid[0].size(),count=0; 
        for(int i =0;i<row;i++)
            for(int j=0;j<col;j++)
                if((i==0 || j==0 || i==row-1 || j==col-1) && grid[i][j]==1)
                dfs(i,j,grid,count,false);
         for(int i =0;i<row;i++)
            for(int j=0;j<col;j++)
                if(grid[i][j]==1){
                    dfs(i,j,grid,count,true);
                    // count++;
                }
        return count;
    }
};

File: 1020-number-of-enclaves/1020-number-of-enclaves.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        def isPossible(x,y):
            return x>=0 and x<len(grid) and y>=0 and y<len(grid[0])
                
        def dfs(x,y):
            if grid[x][y]==0: 
                return
            grid[x][y]=0
            for k1,k2 in [[1,0],[-1,0],[0,1],[0,-1]]:
                new_x,new_y=x+k1,y+k2
                if isPossible(new_x,new_y):
                    dfs(new_x,new_y)
            return 
        
        
        
        cnt=0
        rows,cols=len(grid),len(grid[0])
        
        # print(grid)
        # iterate on the borders & remove them
        for i in range(rows):
            if grid[i][0]==1:
                dfs(i,0)
            if grid[i][cols-1]==1:
                dfs(i,cols-1)
        
        for c in range(cols):
            if grid[0][c]==1:
                dfs(0,c)
            if grid[rows-1][c]==1:
                dfs(rows-1,c)
                
        # print(grid)
        total=sum([ sum(row) for row in grid])
        
        return total